tf_quant_finance/black_scholes/vanilla_prices_test.py

```diff
@@ -334,6 +334,113 @@ def test_binary_vanilla_consistency_exact(self):
         tf.where(is_call_options, -implied_binary_price, implied_binary_price))
     self.assertArrayNear(implied_binary_price, actual_binary_price, 1e-10)
 
+  def test_barrier_option_1d_puo(self):
+    """Function tests barrier option pricing for scalar input"""
+    spots = 100.0
+    rebates = 3.0
+    expiries = 0.5
+    discount_rates = 0.08
+    continuous_dividends = 0.04
+    exp = self.get_barrier_option_test_vals("puo")
+    strikes = exp[0]
+    barriers = exp[1]
+    price_true = exp[2]
+    is_call_options = exp[3]
+    is_barrier_down = exp[4]
+    is_knock_out = exp[5]
+    volatilities = 0.25
+
+    """
+    3 -> cdi
+    2 -> pdi
+    1 -> cui
+    0 -> pui
+    7 -> cdo
+    6 -> pdo
+    5 -> cuo
+    4 -> puo
+    """
+    price = tff.black_scholes.barrier_price(
+        volatilities=volatilities, strikes=strikes,
+        expiries=expiries, spots=spots,
+        discount_rates=discount_rates,
+        continuous_dividends=continuous_dividends,
+        barriers=barriers, rebates=rebates,
+        is_barrier_down=is_barrier_down,
+        is_knock_out=is_knock_out,
+        is_call_options=is_call_options)
+    self.assertAllClose(price, price_true, 10e-3)
+
+  def get_barrier_option_test_vals(self, param):
+    """Function returns testing vals for type of option"""
+    if param == "cdo":
+      return (90, 95, 9.0246, True, True, True, 7)
+    if param == "cdi":
+      return (90, 95, 7.7627, True, True, False, 3)
+    if param == "cuo":
+      return (90, 105, 2.6789, True, False, True, 5)
+    if param == "cui":
+      return (90, 105, 14.1112, True, False, False, 1)
+    if param == "pdo":
+      return (90, 95, 2.2798, False, True, True, 6)
+    if param == "puo":
+      return (90, 105, 3.7760, False, True, False, 4)
+    if param == "pdi":
+      return (90, 95, 2.9586, False, False, True, 2)
+    if param == "pui":
+      return (90, 105, 1.4653, False, False, False, 0)
+
+  def test_barrier_option_2d(self):
+    """Function tests barrier option pricing for vector inputs"""
+    spots = [100., 100., 100., 100., 100., 100., 100., 100.]
+    rebates = [3., 3., 3., 3., 3., 3., 3., 3.]
+    expiries = [.5, .5, .5, .5, .5, .5, .5, .5]
+    discount_rates = [.08, .08, .08, .08, .08, .08, .08, .08]
+    volatilities = [.25, .25, .25, .25, .25, .25, .25, .25]
+    strikes = [90., 90., 90., 90., 90., 90., 90., 90.]
+    barriers = [95., 95., 105., 105., 95., 105., 95., 105.]
+    price_true = [
+        9.024, 7.7627, 2.6789, 14.1112, 2.2798, 3.7760, 2.95586, 1.4653]
```

Comment:
Could you please say where these numbers come from?

Reply:
Added comment to test function!